proposals/NNNN-isolated-default-values.md

```diff
@@ -0,0 +1,173 @@
+
+```swift
+@MainActor func requiresMainActor() -> Int { ... }
+
+@MainActor func useDefaultClosure(
+  closure: () -> Void = {
+    requiresMainActor()
+  }
+) {}
+```
+
+The above `useDefaultClosure` function has a default argument value that is a closure literal. The closure body calls a `@MainActor`-isolated function synchronously, therefore the closure itself must be `@MainActor` isolated.
+
+Note that the inferred actor isolation of a closure literal can never be an actor instance based on the following two properties:
+
+1. To be isolated to an actor instance, a closure must either have its own (explicit) isolated parameter or capture an isolated parameter from its enclosing context.
+2. Closure literals in default arguments cannot capture values.
+
+### Enforcing default value isolation requirements
+
+#### Default argument values
+
+Isolation requirements for default argument expressions are enforced at the caller. If the caller is not in the required isolation domain, the default argument cannot be used and the argument must be specified explicitly. For example:
+
+```swift
+@MainActor func requiresMainActor() -> Int { ... }
+
+@MainActor func useDefault(value: Int = requiresMainActor()) { ... }
+
+@MainActor func mainActorCaller() {
+  useDefault() // okay
+}
+
+func nonisolatedCaller() async {
+  useDefault() // error
+
+  useDefault(value: await requiresMainActor()) // okay
+}
+```
+
+#### Stored property initial values
+
+Isolation requirements for default initializer expressions for stored properties apply in the body of initializers. If an `init` does not match the isolation of the initializer expression, the initialization of that stored property is not emitted at the beginning of the `init`. Instead, the stored property must be explicitly initialized in the body of the `init`. For example:
+
+```swift
+@MainActor func requiresMainActor() -> Int { ... }
+@AnotherActor func requiresAnotherActor() -> Int { ... }
+
+class C {
+  @MainActor var x1: Int = requiresMainActor()
+  @AnotherActor var x2: Int = requiresAnotherActor()
+
+  nonisolated init() {} // error
+
+  nonisolated init(x1: Int, x2: Int) { // okay
+    self.x1 = x1
+    self.x2 = x2
+  }
+
+  @MainActor init(x2: Int) { // okay
+    // 'self.x1' gets assigned to the default value 'requiresMainActor()'
+    self.x2 = x2
+  }
+}
+```
+
+In the above example, the no-parameter `nonisolated init()` is invalid, because it does not initialize `self.x1` and `self.x2`. Because the default initializer expressions require different actor isolation, those values are not used in the `nonisolated` initializer. The other two initializers are valid.
```

Comment:
Most folks don't put global actor isolation directly on stored properties, so I think it might be hard for them to extrapolate from what's here to how it affects them. This might benefit from an example with a `@MainActor class` and some stored properties that have default values.